crates/gen/src/com_interface.rs

```diff
@@ -138,14 +215,81 @@ impl ComInterface {
             impl #name {
                 #(#methods)*
             }
+            #conversions
         }
     }
 
     pub fn dependencies(&self) -> Vec<winmd::TypeDef> {
         self.methods
             .iter()
             .map(|method| method.signature.dependencies())
+            .chain(self.bases.iter().map(|base| base.dependencies()))
             .flatten()
             .collect()
     }
 }
+
+fn gen_constraint(method: &Method) -> TokenStream {
+    let mut tokens = Vec::new();
+
+    for (position, param) in method.signature.params.iter().enumerate() {
+        if !param.is_input || param.is_array {
+            continue;
+        }
+
+        match &param.kind {
+            TypeKind::IUnknown | TypeKind::Interface(_) => {
+                let name = squote::format_ident!("T{}__", position);
+                let into = param.kind.gen();
+                tokens.push(quote! { #name: ::std::convert::Into<::windows::Param<'a, #into>>, });
+            }
+            _ => {}
+        };
+    }
+
+    if !tokens.is_empty() {
+        tokens.insert(0, quote! { 'a, });
+    }
+
+    TokenStream::from_iter(tokens)
+}
+
+fn gen_params(method: &Method) -> TokenStream {
+    let mut tokens = Vec::new();
+
+    for (position, param) in method.signature.params.iter().enumerate() {
+        let name = format_ident(&param.name);
+
+        match &param.kind {
+            TypeKind::IUnknown | TypeKind::Interface(_) if param.is_input && !param.is_array => {
+                let type_tokens = squote::format_ident!("T{}__", position);
+                tokens.push(quote! { #name: #type_tokens, });
+            }
+            _ => {
+                let type_tokens = param.gen_field();
+                tokens.push(quote! { #name: #type_tokens, });
+            }
+        };
+    }
+
+    TokenStream::from_iter(tokens)
+}
+
+fn gen_abi_args(method: &Method) -> TokenStream {
+    let mut tokens = Vec::new();
+
+    for param in method.signature.params.iter() {
```

Comment:
If you use `map` instead of collecting into a vector you can just pass the iterator to `TokenStream::from_iter` and avoid allocating a vector.

Reply:
Nice, done.